packages/console/src/ds-components/KeyValueInputField/index.tsx

```diff
@@ -0,0 +1,127 @@
+import { type AdminConsoleKey } from '@logto/phrases';
+import { type ReactElement, useCallback } from 'react';
+import { type FieldError } from 'react-hook-form';
+
+import CirclePlus from '@/assets/icons/circle-plus.svg';
+import Minus from '@/assets/icons/minus.svg';
+import Button from '@/ds-components/Button';
+import type DangerousRaw from '@/ds-components/DangerousRaw';
+import FormField from '@/ds-components/FormField';
+import IconButton from '@/ds-components/IconButton';
+import TextInput, { type Props as TextInputProps } from '@/ds-components/TextInput';
+
+import * as styles from './index.module.scss';
+
+type FieldType = {
+  key: string;
+  value: string;
+};
+
+type ErrorType = {
+  [K in keyof FieldType]?: FieldError | string | undefined;
+};
+
+// TextInput props getter
+type InputFieldPropsGetter = {
+  [K in keyof FieldType]: (index: number) => Omit<TextInputProps, 'ref'>;
+};
+
+type Props = {
+  title: AdminConsoleKey | ReactElement<typeof DangerousRaw>;
+  tip?: string;
+  fields: Array<FieldType & { id: string }>; // Id is required to uniquely identify each field
+  errors?: Array<ErrorType | undefined>;
+  getInputFieldProps: InputFieldPropsGetter;
+  onRemove: (index: number) => void;
+  onAppend: (field: FieldType) => void;
+};
+
+/**
+ * UI component for key-value input field.
+ *
+ * This component is used to add multiple key-value pairs.
+ * For most of the cases, it is designed to be used along with react-hook-form.
+ * All the input properties are registered with react-hook-form.
+ *
+ * @param {string} title - The title of the field.
+ * @param {string} [tip] - The tip for the field.
+ * @param {FieldType} fields - The array of key-value pairs. @see {@link https://react-hook-form.com/docs/usefieldarray}
+ * @param {ErrorType[]} [errors] - The array of errors for each field. Accepts both string and FieldError from RHF.
+ * @param {Function} onRemove - The function to remove a field. @see {@link https://react-hook-form.com/docs/usefieldarray}
+ * @param {Function} onAppend - The function to append a new field. @see {@link https://react-hook-form.com/docs/usefieldarray}
+ * @param {InputFieldPropsGetter} getInputFieldProps - The function bundle to get the input field props for each field. e.g. Use React Hook Form's register method to register the input field.
+ */
+function KeyValueInputField({
+  title,
+  tip,
+  fields,
+  errors,
+  getInputFieldProps,
+  onRemove,
+  onAppend,
+}: Props) {
+  const renderErrors = useCallback(
+    (index: number, key: keyof FieldType) => {
+      const error = errors?.[index]?.[key];
+
+      if (!error) {
+        return null;
+      }
+
+      if (typeof error === 'string') {
+        return <div className={styles.error}>{error}</div>;
+      }
+
+      return <div className={styles.error}>{error.message}</div>;
+    },
+    [errors]
+  );
+
+  return (
+    <FormField title={title} tip={tip}>
```

Comment:
We don't need this `FormField` wapper for this component, since we always have some custom content inside the `FormField`. e.g., when we want to add a description for this field, it may be above the input component, or under the input component.

Reply:
Updated